fpx/src/data.rs

```diff
@@ -1,15 +1,80 @@
+use crate::api::models::Request;
 use anyhow::{Context, Result};
-use libsql::{de, params, Builder, Connection, Rows, Transaction};
-use serde::de::DeserializeOwned;
+use libsql::{params, Builder, Connection};
 use std::collections::BTreeMap;
 use std::fmt::Display;
+use std::marker::PhantomData;
+use std::ops::Deref;
 use std::path::{Path, PathBuf};
+use std::sync::Arc;
 use thiserror::Error;
-
-use crate::models::Request;
+use tokio::sync::{Mutex, MutexGuard};
+use tracing::instrument;
+use util::RowsExt;
 
 pub mod migrations;
-mod models;
+pub mod models;
+#[cfg(test)]
+mod tests;
+pub mod util;
+
+pub struct Transaction<'a, T>
+where
+    T: TransactionType,
+{
+    tx: libsql::Transaction,
+    guard: MutexGuard<'a, ()>,
```

Comment:
why not have the mutexguard wrap the transaction?

Reply:
I like that you as a dev are only interacting with a Transaction. If we were to change the implementation further down the line we'd only need to change this file (though most like other usages of the file also won't require any changes).

Reply:
the `MutexGuard` implements `Deref` and `DerefMut` to the inner type so interacting with it will be exactly the same as with a normal transaction, I just think it makes sense to have the guard wrap the type it protects in order to not have them accidently de-couple from each other through some code error

Reply:
If we do wrap it in a MutexGuard, then we are unable to consume the Transaction when we commit or rollback. Since the commit and rollback require a owned transaction, and with a Mutex guard we are unable to get this, we can only get a ref or mut ref.

We are using the mutex as a way of controlling concurrency, not to prevent multiple access to the same data.